### PR TITLE
Start figuring out which things we need to rebuild in fontc

### PR DESCRIPTION
**Pull request by rsheeter**
Rename depgraph to FileStateSet because that seems to better reflect how we use it.

Builds on #11, merge that first for a smaller diff.

### CODE REVIEW

**Comment by anthrotype** [comment]
> Rename depgraph to FileStateSet

I wish this was a distinct commit so it was easier to disentangle with other changes in this PR

**Review by cmyr — APPROVED** [review]
A bunch of style nits and notes but the overall logic seems reasonable!

File: fontir/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+use std::{error::Error, io, path::PathBuf};
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum FontIrError {
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/error.rs:6]: But I used to Java and we UseVeryLongANdDefinitelyHelpfulNamesALotFactoryFactory!

[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/error.rs:6]: (fixing)

File: fontir/src/filestate.rs [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+                        self.entries
+                            .insert(new_path.clone(), FileState::of(&new_path)?);
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn new_files(&self, dirs_visited: &mut HashSet<PathBuf>, dir: &Path) -> Option<Vec<PathBuf>> {
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/filestate.rs:138]: An empty vec *doesn't allocate*? At all?

File: ufo2fontir/src/source.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            // Track files within each UFO
+            // The UFO dir *must* exist since we were able to find fontinfo in it earlier
+            let ufo_dir = designspace_dir.join(source.filename);
+
+            let glif_files = glif_files(&ufo_dir)?;
+            let glyph_dir = ufo_dir.join("glyphs");
+            if !glyph_dir.is_dir() {
+                return Err(FontIrError::DirectoryExpected(glyph_dir));
+            }
+            for (glyph_name, glif_file) in glif_files.into_iter() {
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply ufo2fontir/src/source.rs:70]: Fixed, ty

File: fontir/src/filestate.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/filestate.rs:138]: Fixed, ty

File: fontc/src/main.rs [hunk 4]
```diff
@@ -1,3 +1,287 @@
...
+            warn!("Unable to parse prior config {:#?}", prior_config);
+            return true;
+        }
+        let prior_config = prior_config.unwrap();
+        *self != prior_config
+    }
+}
+
+fn establish_build_dir() -> PathBuf {
+    let build_dir = PathBuf::from("build");
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontc/src/main.rs:56]: I do need a pathbuf in the end but I can return it as to_path_buf

File: fontc/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+use std::{io, path::PathBuf};
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum FontcError {
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontc/src/error.rs:6]: Done, ty

File: fontc/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1 @@
+pub mod error;
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontc/src/lib.rs:1]: Added, ty

[the remark this answers is not in the document]
Comment (rsheeter) [reply fontc/src/lib.rs:1]: Fixed

File: fontir/src/filestate.rs [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+    entries: HashMap<PathBuf, FileState>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+struct FileState {
+    mtime: FileTime,
+    size: u64,
+}
+
+// We use a named field because toml doesn't like tuple-structs very much
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/filestate.rs:24]: Good idea, will do.

**Comment by rsheeter** [comment]
> I wish this was a distinct commit so it was easier to disentangle with other changes in this PR

In retrospect I agree.

**Pull request merged**